certcache.go

```diff
@@ -16,25 +16,102 @@ package amppackager
 
 import (
 	"bytes"
+	"context"
 	"crypto/x509"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	"log"
 	"net/http"
+	"net/url"
+	"sync"
 	"time"
 
 	"github.com/WICG/webpackage/go/signedexchange/certurl"
 	"github.com/julienschmidt/httprouter"
+	"github.com/pkg/errors"
+	"github.com/pquerna/cachecontrol"
+	"golang.org/x/crypto/ocsp"
 )
 
+// The maximum representable time, per https://stackoverflow.com/questions/25065055/what-is-the-maximum-time-time-in-go.
+var infiniteFuture = time.Unix(1<<63-62135596801, 999999999)
+
+// The OCSP code below aims to satisfy items 1-8 of the "sleevi" doc:
+// https://gist.github.com/sleevi/5efe9ef98961ecfb4da8
+// Item #9 should be added. Item #10 is N/A for SXGs.
+
+// 1MB is the maximum used by github.com/xenolf/lego/acmev2 in GetOCSPForCert.
+// Alternatively, here's a random documented example of a 20K limit:
+// https://www.ibm.com/support/knowledgecenter/en/SSPREK_9.0.0/com.ibm.isam.doc/wrp_stza_ref/reference/ref_ocsp_max_size.html
+const maxOCSPResponseBytes = 1024 * 1024
+
+// How often to check if OCSP stapling needs updating.
+const ocspCheckInterval = 1 * time.Hour
+
 type CertCache struct {
-	// TODO(twifkak): Support multiple certs.
+	// TODO(twifkak): Support multiple cert chains (for different domains, for different roots).
 	certName string
-	cert     *x509.Certificate
+	certs    []*x509.Certificate
+	// TODO(twifkak): Extract an ocsp struct and a transactional in-memory thingie.
+	// A lock over ocsp and ocspUpdateAfter. This is not technically
+	// transactional (since the upgrade from read lock to write lock is
+	// non-atomic), but good enough, given that the write only happens once
+	// an hour at most.
+	ocspMu          sync.RWMutex
+	ocsp            []byte
+	ocspUpdateAfter time.Time
+	// TODO(twifkak): Implement a registry of Updateable instances which can be configured in the toml.
+	ocspFile Updateable
+	client   http.Client
+
+	// Fakes exposed for testing. Should be zero in production.
+	fakeOCSPServer string
```

Comment:
How about creating a func assigned to a var that can be overridden in your test?

in real code:
var getOCSPServerOrDefault = func(certs []x509.Certificate, orig []byte) {
    if len(this.certs[0].OCSPServer) < 1 {
        log.Println("Cert missing OCSPServer.")
￼ 	return orig
    } else {
￼ 	// This is a URI, per https://tools.ietf.org/html/rfc5280#section-4.2.2.1.
￼  	return this.certs[0].OCSPServer[0]
    }
}

In test:
var getOCSPServer = func(_ []*x509.Certificate, _ []byte) {
  return fake
}

Reply:
Done.